drivers/adc/ad7616/iio_ad7616.c

```diff
@@ -0,0 +1,383 @@
+*******************************************************************************/
+#ifdef IIO_SUPPORT
+
+#include <stdio.h>
+#include <math.h>
+
+#include "ad7616.h"
+#include "iio_ad7616.h"
+#include "iio.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_error.h"
+
+#define AD7616_CH(_idx) {\
+	.name = "voltage" #_idx, \
+	.ch_type = IIO_VOLTAGE,\
+	.ch_out = IIO_DIRECTION_INPUT,\
+	.indexed = true,\
+	.channel = _idx,\
+	.scan_index = _idx,\
+	.scan_type = &ad7616_iio_scan_type,\
+	.attributes = ad7616_iio_ch_attributes,\
+}
+
+#define AD7616_CHN_ATTR(_name, _priv) {\
+	.name = _name,\
+	.priv = _priv,\
+	.show = iio_ad7616_attr_get,\
+	.store = iio_ad7616_attr_set,\
+}
+
+#define CHN_STORAGE_BITS 16
+/* 16 IIO channel */
+#define CHANNEL_NUMBER 16
+#define RESULTS_NUMBER CHANNEL_NUMBER
+/* Each AD7616 channel has 8 sources */
+#define CHANNEL_SOURCES_NUMBER 8
+
+struct ad7616_iio {
+	struct ad7616_dev *drv_dev;
+	struct iio_device *iio_dev;
+};
+
+enum ad7616_iio_attributes_id {
+	CRC_ATTR_ID,
+	OSR_ATTR_ID,
+	OSR_AVAILABLE_ID,
+};
+
+/**
+ * @brief	Check if the channel number correspond to the channel A of AD7616
+ * @param	channel_id - Id of the channel
+ * @returns	1 if true, 0 otherwise
+ */
+static int32_t is_channel_a(uint32_t channel_id)
+{
+	return channel_id < CHANNEL_SOURCES_NUMBER;
+}
+
+/**
+ * @brief       Read the debug register value
+ * @param       dev[in, out]- Pointer to IIO device instance
+ * @param       reg[in]- Register address to read from
+ * @param       readval[out]- Pointer to variable to read data into
+ * @return      0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad7616_debug_reg_read(void *device, uint32_t reg,
+		uint32_t *val)
+{
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	return ad7616_read(dev,  (uint8_t) reg, (uint16_t *)val);
+}
+
+/**
+ * @brief       Write the debug register value
+ * @param       dev[in, out]- Pointer to IIO device instance
+ * @param       reg[in]- Register address to read from
+ * @param       writeval[out]- Pointer to variable to write data into
+ * @return      0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad7616_debug_reg_write(void *device, uint32_t reg,
+		uint32_t val)
+{
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	return ad7616_write(dev, (uint8_t)reg, (uint16_t)val);
+}
+
+static int get_raw(void *device,
+		   char *buf,
+		   uint32_t len,
+		   const struct iio_ch_info *channel,
+		   intptr_t id)
+{
+	int ret;
+	uint8_t index = is_channel_a(channel->ch_num) ? channel->ch_num :
+			channel->ch_num - CHANNEL_SOURCES_NUMBER;
+	struct ad7616_conversion_result res[RESULTS_NUMBER];
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	ret = ad7616_read_data_serial(dev, res, 1);
+	if (ret)
+		return ret;
+
+	if (is_channel_a(channel->ch_num))
```

Comment:
is_channel_a is called twice for the same variable. it would be better to store the result of the first call and use it here, without calling the function again